### ISSUE TITLE
Copy features from BidOnAddresses.sol

### ISSUE THREAD

**Issue by vporton** [gnosis/conditional-tokens-contracts#112]
[This contract](https://github.com/vporton/conditional-tokens-contracts/blob/bid-addresses/contracts/BidOnAddresses.sol) has some features that `ConditionalTokens.sol` should but doesn't support.

Please copy some (all reasonably "copyable") features from `BidOnAddresses.sol` to `ConditionalTokens.sol`.

For example, ERC-1155 collaterals would be useful. (We can wrap ERC-20 inside a ERC-1155 wrapper; well, this wrapper wouldn't be fully compatible to ERC-1155 because we can't emit all the events, but that will work for us nevertheless.)

**Comment by cag** [comment]
What features specifically do you want to see copied from that contract? Maybe we can break these features down into separate issues, and at some point a new version of the conditional tokens may be implemented.

The problem is diverting too much energy to getting folks to migrate to a new version before the value is there, which is why this repo is just collecting issues. Smart contract development is quite similar to hardware development in that respect...

**Comment by vporton** [comment]
> What features specifically do you want to see copied from that contract? Maybe we can break these features down into separate issues, and at some point a new version of the conditional tokens may be implemented.
Yes, need to break into separate issues, but I don't plan now to work on this idea thoroughly. I suggest you to do this.

One of of the features is NxM combinations where N is the number of oracles and M is the number of markets (sets of conditional tokens).

Another is using ERC-1155 for both conditionals and collaterals.

There are more features that can be "copied".

Such features require a new version of your contract with a new API.
> The problem is diverting too much energy to getting folks to migrate to a new version before the value is there, which is why this repo is just collecting issues. Smart contract development is quite similar to hardware development in that respect...

I realize that it requires work. But in some future it should be done. Well, I myself don't use it.

**Comment by cag** [comment]
Just for future reference, I think https://github.com/gnosis/conditional-tokens-contracts/pull/110 is related to this issue?

**Comment by vporton** [comment]
> Just for future reference, I think #110 is related to this issue?

Yes, but take note that there is also https://github.com/vporton/conditional-tokens-contracts/tree/bid-addresses-solidity-0.7

**Comment by cag** [comment]
> One of of the features is NxM combinations where N is the number of oracles and M is the number of markets (sets of conditional tokens).

Forgive me for the late response. I've been trying to grok what's been written. From my current understanding, "markets" here actually means something analogous to "conditions," and NxM combinations is already possible with the current contract, even without any transactions, so it's even cheaper since these combinations exist counterfactually.

> Another is using ERC-1155 for both conditionals and collaterals.

Yes, we're aware of this: https://github.com/gnosis/conditional-tokens-contracts/issues/56

**Comment by vporton** [comment]
> > One of of the features is NxM combinations where N is the number of oracles and M is the number of markets (sets of conditional tokens).
> 
> Forgive me for the late response. I've been trying to grok what's been written. From my current understanding, "markets" here actually means something analogous to "conditions," and NxM combinations is already possible with the current contract, even without any transactions, so it's even cheaper since these combinations exist counterfactually.

No, there are 2 or more conditions per market.
